### PR TITLE
[Backend Configuration IIIb] Configure DynamicTable router for set_data_io

### PR DESCRIPTION
**Pull request by CodyCBakerPhD**
Adds support for dynamic tables to the backend configuration tool, which previously had a bug related to https://github.com/hdmf-dev/hdmf/pull/1013

@alejoe91 Do note that when testing this, you will need to use HDMF dev (`pip install git+https://github.com/hdmf-dev/hdmf.git@dev` or fresh `pip install -e` on this branch)

### CODE REVIEW

File: tests/test_minimal/test_tools/test_backend_and_dataset_configuration/test_helpers/test_configure_backend_defaults.py
Comment (CodyCBakerPhD) [inline tests/test_minimal/test_tools/test_backend_and_dataset_configuration/test_helpers/test_configure_backend_defaults.py:69]:
```suggestion
```

File: tests/test_minimal/test_tools/test_backend_and_dataset_configuration/test_helpers/test_configure_backend_overrides.py
Comment (CodyCBakerPhD) [inline tests/test_minimal/test_tools/test_backend_and_dataset_configuration/test_helpers/test_configure_backend_overrides.py:85]:
```suggestion
```

File: tests/test_minimal/test_tools/test_backend_and_dataset_configuration/test_helpers/test_configure_backend_defaults.py
Comment (CodyCBakerPhD) [inline tests/test_minimal/test_tools/test_backend_and_dataset_configuration/test_helpers/test_configure_backend_defaults.py:63]:
```suggestion
    tmpdir: Path, case_name: str, backend: Literal["hdf5", "zarr"]
```

File: tests/test_minimal/test_tools/test_backend_and_dataset_configuration/test_helpers/test_configure_backend_overrides.py
Comment (CodyCBakerPhD) [inline tests/test_minimal/test_tools/test_backend_and_dataset_configuration/test_helpers/test_configure_backend_overrides.py:79]:
```suggestion
    tmpdir: Path, case_name: str, backend: Literal["hdf5", "zarr"]
```

File: tests/test_minimal/test_tools/test_backend_and_dataset_configuration/test_helpers/test_configure_backend_overrides.py
Comment (CodyCBakerPhD) [inline tests/test_minimal/test_tools/test_backend_and_dataset_configuration/test_helpers/test_configure_backend_overrides.py:93]:
```suggestion
    smaller_chunk_shape = (30_000, 64)
    dataset_configuration.chunk_shape = smaller_chunk_shape
```

File: tests/test_minimal/test_tools/test_backend_and_dataset_configuration/test_helpers/test_configure_backend_defaults.py
Comment (alejoe91) [inline tests/test_minimal/test_tools/test_backend_and_dataset_configuration/test_helpers/test_configure_backend_defaults.py:62]: @CodyCBakerPhD you should remove or specify `case_name` here!
```suggestion
def test_simple_dynamic_table(tmpdir: Path, backend: Literal["hdf5", "zarr"]):
```

File: tests/test_minimal/test_tools/test_backend_and_dataset_configuration/test_helpers/test_configure_backend_defaults.py
Comment (alejoe91) [inline tests/test_minimal/test_tools/test_backend_and_dataset_configuration/test_helpers/test_configure_backend_defaults.py:75]:
```suggestion
    nwbfile_path = str(tmpdir / f"test_configure_defaults_dynamic_table.nwb.{backend}")
```
ditto

File: tests/test_minimal/test_tools/test_backend_and_dataset_configuration/test_helpers/test_configure_backend_overrides.py
Comment (CodyCBakerPhD) [inline tests/test_minimal/test_tools/test_backend_and_dataset_configuration/test_helpers/test_configure_backend_overrides.py:78]:
```suggestion
def test_simple_dynamic_table_override(tmpdir: Path, backend: Literal["hdf5", "zarr"]):
```

File: tests/test_minimal/test_tools/test_backend_and_dataset_configuration/test_helpers/test_configure_backend_overrides.py
Comment (CodyCBakerPhD) [inline tests/test_minimal/test_tools/test_backend_and_dataset_configuration/test_helpers/test_configure_backend_overrides.py:102]:
```suggestion
```

**Comment by alejoe91** [comment]
Testing now @CodyCBakerPhD !!! :)

File: requirements-minimal.txt
Comment (h-mayorquin) [inline requirements-minimal.txt:6]: Aren't we waiting for this to be merged?

File: src/neuroconv/tools/nwb_helpers/_configure_backend.py
Comment (h-mayorquin) [inline src/neuroconv/tools/nwb_helpers/_configure_backend.py:25]: I have make this point before but I don't remember you answer, why do you favor in-lining definitions so much? 

It does get in the way of my debugging workflow. That's not a strong argument but if there is no argument in favor either could we do something like this instead?
```
        # TODO: update buffer shape in iterator, if present
        neurodata_object = nwbfile_objects[object_id]
        if isinstance(neurodata_object, Data):
            neurodata_object.set_data_io(data_io_class=data_io_class, data_io_kwargs=data_io_kwargs)
        elif isinstance(neurodata_object, TimeSeries):
            neurodata_object.set_data_io(
                dataset_name=dataset_name, data_io_class=data_io_class, **data_io_kwargs
            )
        else:  # Strictly speaking, it would be odd if a backend_configuration led to this, but might as well be safe
            raise NotImplementedError(
                f"Unsupported object type {type(neurodata_object)} for backend "
                f"configuration of {neurodata_object.name}!"
            )
```

File: src/neuroconv/tools/nwb_helpers/_configure_backend.py
Comment (h-mayorquin) [inline src/neuroconv/tools/nwb_helpers/_configure_backend.py:27]: What happens with the timestamps if they are present?  Are they chunked compressed?

File: tests/test_minimal/test_tools/test_backend_and_dataset_configuration/test_helpers/test_configure_backend_overrides.py
Comment (h-mayorquin) [inline tests/test_minimal/test_tools/test_backend_and_dataset_configuration/test_helpers/test_configure_backend_overrides.py:94]: This if is unecessary.

File: tests/test_minimal/test_tools/test_backend_and_dataset_configuration/test_helpers/test_configure_backend_defaults.py
Comment (h-mayorquin) [inline tests/test_minimal/test_tools/test_backend_and_dataset_configuration/test_helpers/test_configure_backend_defaults.py:76]: I would also avoid inlining the IO class here. I think it makes the code harder to read but is marginal.

File: tests/test_minimal/test_tools/test_backend_and_dataset_configuration/test_helpers/test_configure_backend_overrides.py
Comment (h-mayorquin) [reply tests/test_minimal/test_tools/test_backend_and_dataset_configuration/test_helpers/test_configure_backend_overrides.py:94]: Oh, I remember you had this in another PR. I think you like it like this? I guess the code feels more extensible / general this way even if it not used. Well, you won't find me arguing for reducing lines as a virtue so I guess this is fine.

File: tests/test_minimal/test_tools/test_backend_and_dataset_configuration/test_helpers/test_configure_backend_defaults.py
Comment (h-mayorquin) [inline tests/test_minimal/test_tools/test_backend_and_dataset_configuration/test_helpers/test_configure_backend_defaults.py:1]: Maybe we could add a ragged array tests. Does chunking makes sense on those cases? I guess it is very difficult to define a good chunking pattern unlesss your entries are homogeneous.

**Review by h-mayorquin — COMMENTED** [review]
This looks good to me. Some small comments but the bigger issue is, are we pinning to a hdmf dev branch? What's the justication of this? is this needed with urgency?

File: requirements-minimal.txt
Comment (CodyCBakerPhD) [reply requirements-minimal.txt:6]: I assume you mean 'released' - I will poke the HDMF team to get one out

It's for https://github.com/hdmf-dev/hdmf/pull/1013, the essential fix to even make it possible to set the data IO for table columns

File: src/neuroconv/tools/nwb_helpers/_configure_backend.py
Comment (CodyCBakerPhD) [reply src/neuroconv/tools/nwb_helpers/_configure_backend.py:25]: It simply does not interfere with my debugging workflow, nor does it alter readability from my perspective when it's something so small (literally just fetching a value from a dict in this case)

I'm fine with doing it your way if you comment on it, but in general IDK if I will ever make it as hyper focused of a habit as you do

File: src/neuroconv/tools/nwb_helpers/_configure_backend.py
Comment (CodyCBakerPhD) [reply src/neuroconv/tools/nwb_helpers/_configure_backend.py:27]: Ought to be

File: tests/test_minimal/test_tools/test_backend_and_dataset_configuration/test_helpers/test_configure_backend_overrides.py
Comment (CodyCBakerPhD) [reply tests/test_minimal/test_tools/test_backend_and_dataset_configuration/test_helpers/test_configure_backend_overrides.py:94]: Yes, this 4 line block is kind of silly but important to note how specific it is to GZIP syntax; if/when more tests are added for other compressors they will not look like this

Granted this is also something that was separated into different functions before the combination with parametrize, where each individual function would have been the extensible thing

So the if/else stuff here just because it's all together in one function

File: tests/test_minimal/test_tools/test_backend_and_dataset_configuration/test_helpers/test_configure_backend_defaults.py
Comment (CodyCBakerPhD) [reply tests/test_minimal/test_tools/test_backend_and_dataset_configuration/test_helpers/test_configure_backend_defaults.py:1]: It would; I think we do have general configuration tests for this, just not for the `configure_backend` call since in principle there's nothing at all different about those special columns that have `_index` in the name (they are still just VectorData at the end of the day)

File: tests/test_minimal/test_tools/test_backend_and_dataset_configuration/test_helpers/test_configure_backend_defaults.py
Comment (CodyCBakerPhD) [reply tests/test_minimal/test_tools/test_backend_and_dataset_configuration/test_helpers/test_configure_backend_defaults.py:76]: Sure

**Comment by CodyCBakerPhD** [comment]
> What's the justication of this? is this needed with urgency?

The urgency is this is a bug Alessio found when trying out the tools; unable to write something on the electrodes table

Granted there are other ways to disable that detection (on our side, or the users), but it's an intended thing for this feature to support so we should fix it as quickly as we can

File: requirements-minimal.txt
Comment (h-mayorquin) [reply requirements-minimal.txt:6]: Yes, correct, I meant released. All right.

File: tests/test_minimal/test_tools/test_backend_and_dataset_configuration/test_helpers/test_configure_backend_overrides.py
Comment (h-mayorquin) [reply tests/test_minimal/test_tools/test_backend_and_dataset_configuration/test_helpers/test_configure_backend_overrides.py:94]: Sure, you are keeping it as a template for how the tests should be. Fair enough.

File: src/neuroconv/tools/nwb_helpers/_configure_backend.py
Comment (h-mayorquin) [reply src/neuroconv/tools/nwb_helpers/_configure_backend.py:27]: Are they `Data` or a `TimeSeries`? I guess that was my question.

**Comment by h-mayorquin** [comment]
> The urgency is this is a bug Alessio found when trying out the tools; unable to write something on the electrodes table

All right, fine by me to merge. I am not that concerned about pinning to a specific branch or commit outside of a release.

I guess that @alejoe91 will test this so he can approve it once it works.

File: src/neuroconv/tools/nwb_helpers/_configure_backend.py
Comment (CodyCBakerPhD) [reply src/neuroconv/tools/nwb_helpers/_configure_backend.py:27]: They are a dataset of the TimeSeries

Comment (CodyCBakerPhD) [reply src/neuroconv/tools/nwb_helpers/_configure_backend.py:27]: (it would enter line 29 here with `dataset_name="timestamps"`)

File: src/neuroconv/tools/nwb_helpers/_configure_backend.py
Comment (CodyCBakerPhD) [inline src/neuroconv/tools/nwb_helpers/_configure_backend.py:35]:
```suggestion
       nwbfile_object = nwbfile_objects[object_id]
        if isinstance(nwbfile_object, Data):
            nwbfile_object.set_data_io(data_io_class=data_io_class, data_io_kwargs=data_io_kwargs)
        elif isinstance(nwbfile_object, TimeSeries):
            nwbfile_object.set_data_io(
                dataset_name=dataset_name, data_io_class=data_io_class, **data_io_kwargs
            )
        else:  # Strictly speaking, it would be odd if a backend_configuration led to this, but might as well be safe
            raise NotImplementedError(
                f"Unsupported object type {type(nwbfile_object)} for backend "
                f"configuration of {nwbfile_object.name}!"
            )
```

File: tests/test_minimal/test_tools/test_backend_and_dataset_configuration/test_helpers/test_configure_backend_defaults.py
Comment (CodyCBakerPhD) [inline tests/test_minimal/test_tools/test_backend_and_dataset_configuration/test_helpers/test_configure_backend_defaults.py:79]:
```suggestion
    NWB_IO = BACKEND_NWB_IO[backend]
    with NWB_IO(path=nwbfile_path, mode="w") as io:
        io.write(nwbfile)

    with NWB_IO(path=nwbfile_path, mode="r") as io:
```

File: tests/test_minimal/test_tools/test_backend_and_dataset_configuration/test_helpers/test_configure_backend_overrides.py
Comment (CodyCBakerPhD) [inline tests/test_minimal/test_tools/test_backend_and_dataset_configuration/test_helpers/test_configure_backend_overrides.py:105]: NWB_IO = BACKEND_NWB_IO[backend]
    with NWB_IO(path=nwbfile_path, mode="w") as io:
        io.write(nwbfile)

    with NWB_IO(path=nwbfile_path, mode="r") as io:

File: src/neuroconv/tools/nwb_helpers/_configure_backend.py
Comment (CodyCBakerPhD) [inline src/neuroconv/tools/nwb_helpers/_configure_backend.py:36]:
```suggestion
        nwbfile_object = nwbfile_objects[object_id]
        if isinstance(nwbfile_object, Data):
            nwbfile_object.set_data_io(data_io_class=data_io_class, data_io_kwargs=data_io_kwargs)
        elif isinstance(nwbfile_object, TimeSeries):
            nwbfile_object.set_data_io(
                dataset_name=dataset_name, data_io_class=data_io_class, **data_io_kwargs
            )
        else:  # Strictly speaking, it would be odd if a backend_configuration led to this, but might as well be safe
            raise NotImplementedError(
                f"Unsupported object type {type(nwbfile_object)} for backend "
                f"configuration of {nwbfile_object.name}!"
            )
```

**Comment by CodyCBakerPhD** [comment]
@alejoe91 If you can confirm that this allows you to proceed testing `configure_backend` around that electrodes table issue (but separate from the writing zeros issue, which will be fixed in a follow-up)

**Comment by alejoe91** [comment]
> @alejoe91 If you can confirm that this allows you to proceed testing `configure_backend` around that electrodes table issue (but separate from the writing zeros issue, which will be fixed in a follow-up)

confirmed

**Comment by CodyCBakerPhD** [comment]
> confirmed

@alejoe91 Sorry, 'confirm by approving' as  per https://github.com/catalystneuro/neuroconv/pull/700#issuecomment-1877376978

**Review by alejoe91 — APPROVED** [review]

**Comment by h-mayorquin** [comment]
Thanks for working on this @CodyCBakerPhD. And lots of love for dealing with questions as usual!

**Pull request merged**
